aws/resource_aws_servicecatalog_product.go

```diff
@@ -0,0 +1,450 @@
+	}
+	input.IdempotencyToken = aws.String(resource.UniqueId())
+	input.SetProvisioningArtifactParameters(&artifactProperties)
+	log.Printf("[DEBUG] Creating Service Catalog Product: %s %s", input, artifactProperties)
+
+	resp, err := conn.CreateProduct(&input)
+	if err != nil {
+		return fmt.Errorf("creating ServiceCatalog product failed: %s", err)
+	}
+	d.SetId(*resp.ProductViewDetail.ProductViewSummary.ProductId)
+	if err := waitForServiceCatalogProductStatus(conn, d); err != nil {
+		return err
+	}
+	return resourceAwsServiceCatalogProductRead(d, meta)
+}
+
+func waitForServiceCatalogProductStatus(conn *servicecatalog.ServiceCatalog, d *schema.ResourceData) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{servicecatalog.StatusCreating},
+		// "CREATED" is not documented but seems to be the state it goes to
+		Target: []string{servicecatalog.StatusAvailable, "CREATED"},
+		Refresh: func() (interface{}, string, error) {
+			resp, err := conn.DescribeProductAsAdmin(&servicecatalog.DescribeProductAsAdminInput{
+				Id: aws.String(d.Id()),
+			})
+			if err != nil {
+				return nil, "", err
+			}
+			return resp, aws.StringValue(resp.ProductViewDetail.Status), nil
+		},
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		PollInterval: 3 * time.Second,
+	}
+	_, err := stateConf.WaitForState()
+	return err
+}
+
+func waitForServiceCatalogProductDeletion(conn *servicecatalog.ServiceCatalog, id string) error {
+	stateConf := resource.StateChangeConf{
+		Pending:      []string{servicecatalog.StatusCreating},
+		Target:       []string{""},
+		Timeout:      15 * time.Minute,
+		PollInterval: 3 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+			resp, err := conn.DescribeProductAsAdmin(&servicecatalog.DescribeProductAsAdminInput{
+				Id: aws.String(id),
+			})
+			if err != nil {
+				if isAWSErr(err, servicecatalog.ErrCodeResourceNotFoundException, "") {
+					return 42, "", nil
+				}
+				return 42, "", err
+			}
+
+			return resp, aws.StringValue(resp.ProductViewDetail.Status), nil
+		},
+	}
+	_, err := stateConf.WaitForState()
+	return err
+}
+
+func resourceAwsServiceCatalogProductRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+	log.Printf("[DEBUG] Reading Service Catalog Product with id %s", d.Id())
+	resp, err := conn.DescribeProductAsAdmin(&servicecatalog.DescribeProductAsAdminInput{
+		Id: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, servicecatalog.ErrCodeResourceNotFoundException, "") {
+			log.Printf("[WARN] Service Catalog Product %q not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("reading ServiceCatalog product '%s' failed: %s", d.Id(), err)
+	}
+
+	d.Set("product_arn", resp.ProductViewDetail.ProductARN)
+
+	err = d.Set("tags", tagsToMapServiceCatalog(keyvaluetags.ServicecatalogKeyValueTags(resp.Tags).IgnoreAws().IgnoreConfig(meta.(*AWSClient).IgnoreTagsConfig).Map()))
```

Comment:
see https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/contribution-checklists.md#adding-resource-tagging-support
 for how to handle tagging